### PR TITLE
feat(ui): Implement `<SignUp />` localizations

### PR DESCRIPTION
**Pull request by alexcarpenter**
## Description

- Adds `makeLocalizeable` util
- Implements localizations within `<SignUp />` steps
- Refactors `<Field.Optional />` to be `<Field.Hint />`
- Exposes strings as props on field components

Pulls a lot of existing code from 
- https://github.com/clerk/javascript/blob/aa4cd761585b888706a36a6eed7436a8f0476dbf/packages/clerk-js/src/ui/localization/makeLocalizable.tsx
- https://github.com/clerk/javascript/blob/5764e2911790051589bb5c4f3b1a2c79f7f30c7e/packages/clerk-js/src/ui/localization/localizationKeys.ts

but opted for a function to retrieve the localized string vs the `makeLocalizable` component wrapper.

DEFAULT `enUS`:
Screenshot 2024-06-18 at 3 57 03 PM

CUSTOM `arSA`:
Screenshot 2024-06-18 at 3 56 17 PM

https://linear.app/clerk/issue/SDKI-67/implement-signup-localization

Will follow up with error translations based on this thread https://clerkinc.slack.com/archives/C05ST2R2HM3/p1718741779381069

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [x] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/ui/bundlewatch.config.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "files": [
-    { "path": "./dist/components/sign-in.js", "maxSize": "40kB" },
-    { "path": "./dist/components/sign-up.js", "maxSize": "40kB" }
+    { "path": "./dist/components/sign-in.js", "maxSize": "80kB" },
+    { "path": "./dist/components/sign-up.js", "maxSize": "80kB" }
```
Comment (alexcarpenter) [inline packages/ui/bundlewatch.config.json:4]: Bumped for now. Localization file bumped over the previous 40kb.

File: packages/ui/src/primitives/field.tsx [hunk 2]
```diff
@@ -65,7 +65,7 @@ export const LabelEnd = React.forwardRef(function Label(
   );
 });
 
-export const Optional = React.forwardRef(function Optional(
+export const Hint = React.forwardRef(function Hint(
```
Comment (alexcarpenter) [inline packages/ui/src/primitives/field.tsx:68]: @joe-bell renamed to `Hint` here to make it useable outside of just using Optional in it.

File: packages/ui/theme-builder/package.json [hunk 3]
```diff
@@ -13,6 +13,7 @@
   "dependencies": {
     "@clerk/nextjs": "file:../../nextjs",
     "@clerk/ui": "file:..",
+    "@clerk/localizations": "file:../../localizations",
```
Comment (alexcarpenter) [inline packages/ui/theme-builder/package.json:16]: Enables pulling in a locale file and providing the the ClerkProvider for testing

```
import { arSA } from '@clerk/localizations';

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <ClerkProvider localization={arSA}>
      <html>
        <body>
          <ThemeBuilder>{children}</ThemeBuilder>
        </body>
      </html>
    </ClerkProvider>
  );
}
```

File: packages/ui/src/common/email-or-phone-number-field.tsx [hunk 4]
```diff
@@ -32,8 +34,7 @@ export function EmailOrPhoneNumberField({
       onChange={setShowPhoneNumberField}
       className={linkButton({ size: 'sm', disabled: props.disabled })}
     >
-      <span className='sr-only'>{toggleDescription}</span>
```
Comment (alexcarpenter) [inline packages/ui/src/common/email-or-phone-number-field.tsx:35]: Removed the hidden text to make sure the button is selectable via the visible text label.

File: packages/ui/src/components/sign-up/sign-up.tsx [hunk 5]
```diff
@@ -119,37 +130,35 @@ function SignUpComponentLoaded() {
...
+                      {emailAddressEnabled ? (
                         <EmailField
+                          label={t('formFieldLabel__emailAddress')}
+                          hintText={t('formFieldHintText__optional')}
                           required={emailAddressRequired}
                           disabled={isGlobalLoading}
                         />
                       ) : null}
 
-                      {phoneNumberEnabled && !emailAddressEnabled ? (
```
Comment (alexcarpenter) [inline packages/ui/src/components/sign-up/sign-up.tsx:134]: SignUp flow does not make use of the toggleable fields only SignIn.

File: packages/ui/bundlewatch.config.json (hunk 1 above)
Comment (joe-bell) [reply packages/ui/bundlewatch.config.json:4]: Ooof that's pretty big. Thinking aloud; what if we reduce the number of localisations by default and make them opt-in instead?

File: packages/ui/src/primitives/field.tsx (hunk 2 above)
Comment (joe-bell) [reply packages/ui/src/primitives/field.tsx:68]: Makes sense to me, good shout

**Comment by joe-bell** [comment]
Really solid work Alex, thank you! I'll go ahead and merge to prevent any conflict pain

**Comment by joe-bell** [comment]
Created two follow-up tickets in Linear; I'll ping you internally

**Review by joe-bell — APPROVED** [review]

**Pull request merged**